[doc] Update external.md (#6869)
taichi-dev/taichi: docs/lang/articles/basic/external.md

There are two ways to import a NumPy array `arr` to the Taichi scope:

> 1. Create a Taichi field `f`, whose shape and dtype match the shape and dtype of `arr`, and call `f.from_numpy(arr)` to copy the data in `arr` into `f`. This approach is preferred when the original array is visited frequently from elsewhere in the Taichi scope (for example, in texture sampling).

Reviewer: ```suggestion
- Create a Taichi field `f`, whose shape and dtype match the shape and dtype of `arr`, and call `f.from_numpy(arr)` to copy the data in `arr` into `f`. This approach is preferred when the original array is visited frequently from elsewhere in the Taichi scope (for example, in texture sampling).
```

2. Pass `arr` as an argument to a kernel or a Taichi function using `ti.types.ndarray()` as type hint. The argument is passed by reference without creating a copy of `arr`. Thus, any modification to this argument from inside a kernel or Taichi function also changes the original array `arr`. This approach is preferred when the kernel or Taichi function that takes in the argument needs to process the original array (for storage or filtering, for example).

:::note
`from_numpy() / from_torch()` can take in any numpy array or torch Tensor, no matter it's contiguous or not. Taichi will manage its own copy of data. However, when passing an argument to a Taichi kernel, only contiguous numpy arrays or torch Tensors are supported.
:::

> ## Data Transfer between NumPy arrays and Taichi Fields

Reviewer: We use the title case for document titles, and the sentence case for subheadings. So only the major words in first-level headings are capitalized,

> ## Data Transfer between NumPy arrays and Taichi Fields

Reviewer: ```suggestion
## Data transfer between NumPy arrays and Taichi fields
```

#       [6, 7, 8]], dtype=int32)
```

> ## Data Transfer between PyTorch/Paddle Tensors and Taichi Fields

Reviewer: ibid.

> ## Data Transfer between PyTorch/Paddle Tensors and Taichi Fields

Reviewer: ```suggestion
## Data transfer between PyTorch/Paddle tensors and Taichi fields
```

    field.from_numpy(array)  # the input array must be of shape (256, 512, 3, 4)
    ```

> 4. When importing data to a struct field, export the data of the corresponding external array as **a dictionary of NumPy arrays, PyTorch tensors, or Paddle tensors** with keys being struct member names and values being struct member arrays. Nested structs are exported as nested dictionaries:

Reviewer: I think we should still use the bullet list here. We can discuss it in our next meeting.

    field.from_numpy(array_dict) # the input array must have the same keys as the field
    ```

> ## Using External Arrays as Taichi Kernel Arguments

Reviewer: ```suggestion
## Using external arrays as Taichi kernel arguments
```

> Use type hint `ti.types.ndarray()` to pass external arrays as kernel arguments. For example:

Reviewer: ```suggestion
Use type hint `ti.types.ndarray()` to pass external arrays as kernel arguments.

### An entry-level example

The following example shows the most basic way to call `ti.types.ndarray()`:
```

> This is an entry-level example to show you how to call `ti.types.ndarray()`. We now illustrate a more advanced usage of this method.

Reviewer: ```suggestion
### Advanced usage
```

> ### Advanced

Reviewer: Should we avoid a single third-level heading in one section? We can discuss it.

> ### Advanced

Reviewer: ```suggestion

```

            b[i, j] = a[i, j] - (a[i-1, j] + a[i, j-1] + a[i+1, j] + a[i, j+1]) / 4
```

> Not only is this code snippet is more readable than the NumPy version above, but it also runs way faster even on the CPU backend.

Reviewer: ```suggestion
Not only is this code snippet more readable than the NumPy version above, but it also runs way faster even on the CPU backend.
```
